emgapi/models.py

```diff
@@ -79,35 +83,147 @@ class Reason(models.IntegerChoices):
+
+            __add_to_suppressibles(
+                descendant_relation.model._meta.object_name,
+                descendants_to_update
+            )
+
+            for descendant in descendants_to_update:
+                for kls, kls_suppressibles in descendant._get_suppression_descendant_tree(
+                        suppressing
+                ).items():
+                    __add_to_suppressibles(
+                        kls,
+                        kls_suppressibles
+                    )
+
+        return suppressibles
+
+    def suppress(self, suppression_reason=None, save=True, propagate=True):
         self.is_suppressed = True
         self.suppressed_at = timezone.now()
         self.suppression_reason = suppression_reason
         if save:
             self.save()
+        if propagate:
+            descendant_tree = self._get_suppression_descendant_tree()
+            for descendants_object_type, descendants in descendant_tree.items():
+                for descendant in descendants:
+                    descendant.is_suppressed = True
+                    descendant.suppression_reason = self.Reason.ANCESTOR_SUPPRESSED
+                m: SuppressibleModel = apps.get_model(app_label='emgapi', model_name=descendants_object_type)
+                m.objects.bulk_update(descendants, fields=['is_suppressed', 'suppression_reason'])
+                logger.info(
+                    f'Propagated suppression of {self} to  {len(descendants)} descendant {descendants_object_type}s'
+                )
         return self
 
-    def unsuppress(self, suppression_reason=None, save=True):
+    def unsuppress(self, save=True, propagate=True):
         self.is_suppressed = False
         self.suppressed_at = None
         self.suppression_reason = None
         if save:
             self.save()
+        if propagate:
+            descendant_tree = self._get_suppression_descendant_tree(suppressing=False)
+            for descendants_object_type, descendants in descendant_tree.items():
+                for descendant in descendants:
+                    descendant.is_suppressed = False
+                    descendant.suppression_reason = None
+                m: SuppressibleModel = apps.get_model(app_label='emgapi', model_name=descendants_object_type)
+                m.objects.bulk_update(descendants, fields=['is_suppressed', 'suppression_reason'])
+                logger.info(
+                    f'Propagated unsuppression of {self} to  {len(descendants)} descendant {descendants_object_type}s'
+                )
         return self
 
     class Meta:
         abstract = True
 
 
-class ENASyncableModel(SuppressibleModel, PrivacyControlledModel):
+class SelectRelatedOnlyQuerySetMixin():
+    def select_related_only(self: QuerySet, related_name: str, only_fields: [str]) -> QuerySet:
+        """
+        Adds a 'select_related' to the queryset, and a 'defer' for all fields other than those specified.
```

Comment:
I'm curious... why "defer" and not just use [`.only()`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#only). Using defer is more flexible as the data will be there, but it may introduce may cause the ORM to make more queries (if my understanding is correct)

Reply:
This is because you can't apply an `.only` just to the fields of a `.select_related`, e.g. you would chain them like:
```python
>>> Run.objects.select_related('sample').only('sample__sample_id')
Traceback (most recent call last):
...
django.core.exceptions.FieldError: Field Run.study cannot be both deferred and traversed using select_related at the same time.
```

(I.e. you'd have to re-add all the fields of the local side of the related to the `only(...)`)

There is probably a way involving subqueries or annotations I suppose, but this use of defer to create a sort of related-scoped `only` seemed a bit easier to reckon with.

If used carefully then no extra queries should be created in the places we use this – it is designed for use in queries where we just need a couple of fields like Accession from the other side of relationship, to render a minimal data into the hyperlinked related relationships.

Reply:
perfect.. I wasn't aware of that caveat of "only()". Thanks :)